### PR TITLE
Implement Network Play with Bug Fix

### PR DESCRIPTION
**Pull request by CalebStadler**
Fixed a checkmate bug that was happening when a king in check could capture a piece to get out of checkmate. Implemented the server and client threads used for network play. Changed the mouse released method in BoardPanel so that communication between server and cliet is possible added in extra functionality with the forfeit and new game buttons in the menu panel for network play.

### CODE REVIEW

**Comment by CalebStadler** [comment]
There was a merge conflict but I got it sorted out.

File: BoardPanel.java [hunk 1]
```diff
@@ -89,95 +94,148 @@ public void mousePressed (MouseEvent e)
...
-                    
-                    //if a pawn is moved to the other side of the board
-                    if(pieces[selected] instanceof Pawn &&
-                        (pieces[selected].getY() == 7 || (pieces[selected].getY() == 0)))
+                    //replace currently selected piece with selected piece type
+                    //if pawn is promoting while playing over network
+                    if(promotionPiece > 0 && !mainFrame.getYourTurn() && 
+                        mainFrame.getConnected())
+                    {
+                        if(promotionPiece == QUEEN)
```
Comment (nwestervelt) [inline BoardPanel.java:129]: Replacing constant with existing constant from PromotionDialog.
```Suggestion
                        if(promotionPiece == PromotionDialog.QUEEN)
```

File: BoardPanel.java [hunk 2]
```diff
@@ -89,95 +94,148 @@ public void mousePressed (MouseEvent e)
...
+                        {
+                            pieces[selected] = new Queen(pieces[selected].getX(), pieces[selected].getY(),
+                                pieces[selected].getPlayer(), mainFrame, pieces);
+                            promotionPiece = NONE;
+                        }
+                        else if(promotionPiece == ROOK)
+                        {
+                            pieces[selected] = new Rook(pieces[selected].getX(), pieces[selected].getY(),
+                                pieces[selected].getPlayer(), mainFrame, pieces);
```
Comment (nwestervelt) [inline BoardPanel.java:138]: Replace constant with existing constant and switch places of Rook and Knight's els if blocks to match below code.
```Suggestion
                        else if(promotionPiece == PromotionDialog.KNIGHT)
                        {
                            pieces[selected] = new Knight(pieces[selected].getX(), pieces[selected].getY(),
                                pieces[selected].getPlayer(), mainFrame, pieces);
```

File: BoardPanel.java [hunk 3]
```diff
@@ -89,95 +94,148 @@ public void mousePressed (MouseEvent e)
...
+                        {
+                            pieces[selected] = new Bishop(pieces[selected].getX(), pieces[selected].getY(),
+                                pieces[selected].getPlayer(), mainFrame, pieces);
+                            promotionPiece = NONE;
+                                
+                        }
+                        else if(promotionPiece == KNIGHT)
+                        {
+                            pieces[selected] = new Knight(pieces[selected].getX(), pieces[selected].getY(),
+                                pieces[selected].getPlayer(), mainFrame, pieces);
```
Comment (nwestervelt) [inline BoardPanel.java:151]: Replace constant with existing constant from PromotionDialog and switch places of Rook and Knight's else if blocks to match below code.
```
                        else if(promotionPiece == PromotionDialog.ROOK)
                        {
                            pieces[selected] = new ROOK(pieces[selected].getX(), pieces[selected].getY(),
                                pieces[selected].getPlayer(), mainFrame, pieces);
```

File: BoardPanel.java [hunk 4]
```diff
@@ -89,95 +94,148 @@ public void mousePressed (MouseEvent e)
...
+                            promotionPiece = NONE;
+                        }
+                        else if(promotionPiece == ROOK)
+                        {
+                            pieces[selected] = new Rook(pieces[selected].getX(), pieces[selected].getY(),
+                                pieces[selected].getPlayer(), mainFrame, pieces);
+                            promotionPiece = NONE;
+                        }
+                        else if(promotionPiece == BISHOP)
```
Comment (nwestervelt) [inline BoardPanel.java:141]: Replace constant with existing constant from PromotionDialog.
```Suggestion
                        else if(promotionPiece == PromotionDialog.BISHOP)
```

File: BoardPanel.java [hunk 5]
```diff
@@ -89,95 +94,148 @@ public void mousePressed (MouseEvent e)
...
                         pd.setVisible(true);
                         PieceAbstract newPiece;
-
-                        //replace currently selected piece with selected piece type
                         if(pd.getSelectedButton() == PromotionDialog.QUEEN)
                         {
                             newPiece = new Queen(pieces[selected].getX(), pieces[selected].getY(),
                                 pieces[selected].getPlayer(), mainFrame, pieces);
                             pieces[selected] = newPiece;
+                            promotionPiece = QUEEN;
```
Comment (nwestervelt) [inline BoardPanel.java:167]: Replace constant with existing constant from PromotionDialog.
```Suggestion
                            promotionPiece = PromotionDialog.QUEEN;
```

File: BoardPanel.java [hunk 6]
```diff
@@ -89,95 +94,148 @@ public void mousePressed (MouseEvent e)
...
                         }
                         else if(pd.getSelectedButton() == PromotionDialog.KNIGHT)
                         {
                             newPiece = new Knight(pieces[selected].getX(), pieces[selected].getY(),
                                 pieces[selected].getPlayer(), mainFrame, pieces);
                             pieces[selected] = newPiece;
+                            promotionPiece = KNIGHT;
```
Comment (nwestervelt) [inline BoardPanel.java:174]: Replace constant with existing constant from PromotionDialog.
```Suggestion
                            promotionPiece = PromotionDialog.KNIGHT;
```

File: BoardPanel.java [hunk 7]
```diff
@@ -89,95 +94,148 @@ public void mousePressed (MouseEvent e)
...
                         }
                         else if(pd.getSelectedButton() == PromotionDialog.BISHOP)
                         {
                             newPiece = new Bishop(pieces[selected].getX(), pieces[selected].getY(),
                                 pieces[selected].getPlayer(), mainFrame, pieces);
                             pieces[selected] = newPiece;
+                            promotionPiece = BISHOP;
```
Comment (nwestervelt) [inline BoardPanel.java:181]: Replace constant with existing constant from PromotionDialog.
```Suggestion
                            promotionPiece = PromotionDialog.BISHOP;
```

File: BoardPanel.java [hunk 8]
```diff
@@ -89,95 +94,148 @@ public void mousePressed (MouseEvent e)
...
                         }
                         else if(pd.getSelectedButton() == PromotionDialog.ROOK)
                         {
                             newPiece = new Rook(pieces[selected].getX(), pieces[selected].getY(),
                                 pieces[selected].getPlayer(), mainFrame, pieces);
                             pieces[selected] = newPiece;
+                            promotionPiece = ROOK;
```
Comment (nwestervelt) [inline BoardPanel.java:188]: Replace constant with existing constant from PromotionDialog.
```Suggestion
                            promotionPiece = PromotionDialog.ROOK;
```

File: BoardPanel.java [hunk 9]
```diff
@@ -89,95 +94,148 @@ public void mousePressed (MouseEvent e)
...
+                    //replace currently selected piece with selected piece type
+                    //if pawn is promoting while playing over network
+                    if(promotionPiece > 0 && !mainFrame.getYourTurn() && 
+                        mainFrame.getConnected())
+                    {
+                        if(promotionPiece == QUEEN)
+                        {
+                            pieces[selected] = new Queen(pieces[selected].getX(), pieces[selected].getY(),
+                                pieces[selected].getPlayer(), mainFrame, pieces);
+                            promotionPiece = NONE;
```
Comment (nwestervelt) [inline BoardPanel.java:133]: Replace constant with 0.
```Suggestion
                            promotionPiece = 0;
```

File: BoardPanel.java [hunk 10]
```diff
@@ -89,95 +94,148 @@ public void mousePressed (MouseEvent e)
...
+                        }
+                        else if(promotionPiece == ROOK)
+                        {
+                            pieces[selected] = new Rook(pieces[selected].getX(), pieces[selected].getY(),
+                                pieces[selected].getPlayer(), mainFrame, pieces);
+                            promotionPiece = NONE;
```
Comment (nwestervelt) [inline BoardPanel.java:139]: Replace constant with 0.
```Suggestion
                            promotionPiece = 0;
```

File: BoardPanel.java [hunk 11]
```diff
@@ -89,95 +94,148 @@ public void mousePressed (MouseEvent e)
...
+                        }
+                        else if(promotionPiece == BISHOP)
+                        {
+                            pieces[selected] = new Bishop(pieces[selected].getX(), pieces[selected].getY(),
+                                pieces[selected].getPlayer(), mainFrame, pieces);
+                            promotionPiece = NONE;
```
Comment (nwestervelt) [inline BoardPanel.java:145]: Replace constant with 0.
                            promotionPiece = 0;
```

File: BoardPanel.java [hunk 12]
```diff
@@ -89,95 +94,148 @@ public void mousePressed (MouseEvent e)
...
+                                
+                        }
+                        else if(promotionPiece == KNIGHT)
+                        {
+                            pieces[selected] = new Knight(pieces[selected].getX(), pieces[selected].getY(),
+                                pieces[selected].getPlayer(), mainFrame, pieces);
+                            promotionPiece = NONE;
```
Comment (nwestervelt) [inline BoardPanel.java:152]: Replace constant with 0.
```Suggestion
                            promotionPiece = 0;
```

File: MainFrame.java [hunk 13]
```diff
@@ -45,12 +62,49 @@ public MainFrame()
         menuPanel = new MenuPanel(this);
         add(menuPanel,BorderLayout.WEST);
 
+        //create meanu, its items, and action handler
```
Comment (nwestervelt) [inline MainFrame.java:65]: Fix typo.
```Suggestion
        //create menu, its items, and action handler
```

File: MainFrame.java [hunk 14]
```diff
@@ -130,6 +184,189 @@ public void addPromotion(PieceAbstract piece)
...
+            {
+                String s=JOptionPane.showInputDialog(MainFrame.this
+                        ,"Enter the port to use",""+port);
+                if(s==null) return;
+                port=Integer.parseInt(s);
+                hostItem.setEnabled(false);
+                joinItem.setEnabled(false);
+                quitItem.setEnabled(true);
+                yourTurn = false;
+                new Server().start();
```
Comment (nwestervelt) [inline MainFrame.java:203]: Add call to the reset method to reset the state of the game, keeping the state between the host and client consistent if one started playing before the other.
```Suggestion
                new Server().start();
                reset();
```

File: MainFrame.java [hunk 15]
```diff
@@ -130,6 +184,189 @@ public void addPromotion(PieceAbstract piece)
...
+                if(s==null) return;
+                host=s;
+                s=JOptionPane.showInputDialog(MainFrame.this
+                        ,"Enter the port to use",""+port);
+                if(s==null) return;
+                port=Integer.parseInt(s);
+                hostItem.setEnabled(false);
+                joinItem.setEnabled(false);
+                quitItem.setEnabled(true);
+                new Client().start();
```
Comment (nwestervelt) [inline MainFrame.java:219]: Add call to the reset method to reset the state of the game, keeping the state between the host and client consistent if one started playing before the other.
```Suggestion
                new Client().start();
                reset();
```

**Review by nwestervelt — APPROVED** [review]
Everything looks fine, and I didn't find any bugs with it while testing it.

**Pull request merged**